src/org/rascalmpl/uri/file/MavenRepositoryURIResolver.java

```diff
@@ -138,92 +122,177 @@ private static String computeMavenCommandName() {
+    }
+
+    @Override
+    public boolean exists(ISourceLocation uri) {
+        try {
+            return reg.exists(resolveInside(uri));
+        }
+        catch (IOException e) {
+            return false;
+        }
+    }
+
+    @Override
+    public long lastModified(ISourceLocation uri) throws IOException {
+        return reg.lastModified(resolveInside(uri));
+    }
+
+    @Override
+    public boolean isDirectory(ISourceLocation uri) {
+        try {
+            return reg.isDirectory(resolveInside(uri));
+        }
+        catch (IOException e) {
+            return false;
+        }
+    }
+
+    @Override
+    public boolean isFile(ISourceLocation uri) {
+        try {
+            return reg.isFile(resolveInside(uri));
+        }
+        catch (IOException e) {
+            return false;
+        }
+    }
+
+    @Override
+    public String[] list(ISourceLocation uri) throws IOException {
+        if (uri.getAuthority().isEmpty() && (uri.getPath().isEmpty() || uri.getPath().equals("/"))) {
+            return listAllMainArtifacts();
+        }
+        else {
+            return reg.listEntries(resolveInside(uri));
+        }
+    }
+
+    /**
+     * Allows browsing through the mvn .m2 repository for "main" artifacts only
+     * @return array of all authorities encoding jar files in the .m2 repo
+     * @throws IOException if the root of the repo is gone or not permitted
+     */
+    private String[] listAllMainArtifacts() throws IOException {
+        try (Stream<Path> files = Files.walk(Paths.get(root.getPath()))) {
+            return files.filter(f -> FileNameUtils.getExtension(f).equals("jar"))
+            .filter(f -> !f.toString().endsWith("-sources"))
+            .filter(f -> !f.toString().endsWith("-javadoc"))
+            .map(f -> {
+                var fl = URIUtil.correctLocation("file", "", f.toString());
+                var parent = URIUtil.getParentLocation(fl);
+                var version = URIUtil.getLocationName(parent);
+                var grandParent = URIUtil.getParentLocation(parent);
+                var artifact = URIUtil.getLocationName(grandParent);
+                var groupId = URIUtil.relativize(root, URIUtil.getParentLocation(grandParent))
+                    .getPath()
+                    .substring(1)
+                    .replaceAll("/", ".");
 
-                // if the path is non-empty we mean to look inside of the jar
-                    if ((!path.isEmpty() && !path.equals("/")) || pathIsJarRoot) {
-                    path = pathIsJarRoot ? "" : path;
-                    // go make a location that points _inside_ of the jar
-                    jarLocation = JarURIResolver.jarify(jarLocation);
-                    jarLocation = URIUtil.getChildLocation(jarLocation, path);
+                if ((artifact + "-" + version + ".jar").equals(URIUtil.getLocationName(fl))) {
+                    return groupId + "!" + artifact + "!" + version;
```

Comment:
The separator should become a constant in this file. As here it's still `!` while other places it's already `--`

Reply:
agreed! thanks for spotting this. The tests should also capture this one.